pruner/src/compiler_pruner.cpp

```diff
@@ -42,16 +44,27 @@ const IR::P4Program *apply_replace_vars(const IR::P4Program *program,
     return program;
 }
 
+void cleanup(std::vector<struct_obj*> *s){
+    for (auto obj : *s)
+    {   
+        delete(obj->fields);   
+        delete(obj);
+    }
+
+    s->clear();
+}
+
+
 const IR::P4Program *apply_unused_decls(const IR::P4Program *program,
                                         P4PRUNER::PrunerConfig pruner_conf) {
     P4::ReferenceMap refMap;
     P4::TypeMap typeMap;
     const IR::P4Program *temp = nullptr;
-
-    PassManager pass_manager({new ExtendedUnusedDeclarations(&refMap)});
-
-    INFO("Applying custom RemoveAllUnusedDeclarations...");
+    std::vector<struct_obj *> used_structs;
+    PassManager pass_manager(
+        {new ExtendedUnusedDeclarations(&refMap, &used_structs)});
     temp = program->apply(pass_manager);
+    cleanup(&used_structs);
```

Comment:
This should not be necessary since the P4C ecosystem uses a garbage collector. 
https://en.wikipedia.org/wiki/Boehm_garbage_collector
So you do not have to worry about allocation and freeing.